core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,406 @@
+                // Reset physical written bytes for rebalanced partitions
+                partitionInfo.resetPhysicalWrittenBytesAtLastRebalance();
+            }
+            else {
+                long writtenBytes = context.estimatePartitionWrittenBytesSinceLastRebalance(new WriterId(writerPartitionId.writerId), rowCount);
+                partitionInfo.addToPhysicalWrittenBytesAtLastRebalance(writtenBytes);
+            }
+        });
+
+        for (int i = 0; i < numberOfWriters; i++) {
+            writerPhysicalWrittenBytesAtLastRebalance.set(i, writerPhysicalWrittenBytes.get(i));
+        }
+    }
+
+    private class RebalanceContext
+    {
+        private final Set<Integer> rebalancedPartitions = new HashSet<>();
+        private final long[] writerPhysicalWrittenBytesSinceLastRebalance;
+        private final long[] writerRowCountSinceLastRebalance;
+        private final long[] writerPriorities;
+        private final List<IndexedPriorityQueue<PartitionIdWithRowCount>> writerMaxPartitions;
+
+        private RebalanceContext(List<Long> writerPhysicalWrittenBytes, Map<WriterPartitionId, Long> partitionsRowCount)
+        {
+            writerPhysicalWrittenBytesSinceLastRebalance = new long[numberOfWriters];
+            writerPriorities = new long[numberOfWriters];
+            for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+                long physicalWrittenBytesSinceLastRebalance =
+                        writerPhysicalWrittenBytes.get(writerId) - writerPhysicalWrittenBytesAtLastRebalance.get(writerId);
+                writerPhysicalWrittenBytesSinceLastRebalance[writerId] = physicalWrittenBytesSinceLastRebalance;
+                writerPriorities[writerId] = physicalWrittenBytesSinceLastRebalance;
+            }
+
+            writerRowCountSinceLastRebalance = new long[numberOfWriters];
+            partitionsRowCount.forEach((writerPartitionId, rowCount) -> writerRowCountSinceLastRebalance[writerPartitionId.writerId] += rowCount);
+
+            writerMaxPartitions = new ArrayList<>(numberOfWriters);
+            for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+                writerMaxPartitions.add(new IndexedPriorityQueue<>());
+            }
+
+            partitionsRowCount.forEach((writerPartitionId, rowCount) ->
+                    writerMaxPartitions
+                            .get(writerPartitionId.writerId)
+                            .addOrUpdate(new PartitionIdWithRowCount(writerPartitionId.partitionId, rowCount), rowCount));
+        }
+
+        private boolean rebalancePartition(WriterId from, WriterId to)
+        {
+            IndexedPriorityQueue<PartitionIdWithRowCount> maxPartitions = writerMaxPartitions.get(from.id);
+
+            while (true) {
+                // Find the partition with maximum written bytes since last rebalance
+                PartitionIdWithRowCount partitionToRebalance = maxPartitions.poll();
```

Comment:
Why not use iterator from IndexedPriorityQueue instead of while + poll and `null` check?

Reply:
poll also removes the element from the queue and index which is important.

Reply:
I've understood what `poll` does, but could you elaborate why index is important?

Reply:
Index is not really important in our case. It's just that we are trying to replicate ` io.trino.execution.scheduler.UniformNodeSelector#equateDistribution` which seems easy to understand and based on `IndexedPriorityQueue`. We could've also used `TreeMap` or built-in `PriorityQueue`. However still, we will have to do an extra call to remove an element from the data structure. Besides, I think it's not really a big deal.

If you think it's better we can remove the null check like this:

```
while (!maxPartitions.isEmpty()) {
      // Find the partition with maximum written bytes since last rebalance
      PartitionIdWithRowCount partitionToRebalance = maxPartitions.poll();
      ...
}
```

Also, take a look at this comment: https://github.com/trinodb/trino/pull/14718#discussion_r1012697894